lib/matplotlib/backends/backend_qt.py

```diff
@@ -682,6 +682,115 @@ def set_window_title(self, title):
         self.window.setWindowTitle(title)
 
 
+class _IconEngine(QtGui.QIconEngine):
+    """
+    Custom QIconEngine that automatically handles DPI scaling for tools icons.
+
+    This engine provides icons on-demand with proper scaling based on the current
+    device pixel ratio, eliminating the need for manual refresh when DPI changes.
+    """
+
+    def __init__(self, image_path, toolbar=None):
+        super().__init__()
+        self.image_path = image_path
+        self.toolbar = toolbar
+
+    def _is_dark_mode(self):
+        return self.toolbar.palette().color(self.toolbar.backgroundRole()).value() < 128
+
+    def paint(self, painter, rect, mode, state):
+        """Paint the icon at the requested size and state."""
+        pixmap = self.pixmap(rect.size(), mode, state)
+        if not pixmap.isNull():
+            painter.drawPixmap(rect, pixmap)
+
+    def pixmap(self, size, mode, state):
+        """Generate a pixmap for the requested size, mode, and state."""
+        if size.width() <= 0 or size.height() <= 0:
+            return QtGui.QPixmap()
+
+        # Get the current device pixel ratio
+        if self.toolbar:
+            dpr = self.toolbar.devicePixelRatioF() or 1
+        else:
+            dpr = 1
+
+        # Try SVG first, then fall back to PNG
+        svg_path = self.image_path.with_suffix('.svg')
+        if svg_path.exists():
+            return self._create_pixmap(svg_path, size, dpr, is_svg=True)
+        else:
+            large_path = self.image_path.with_name(self.image_path.stem + '_large.png')
+            path = large_path if large_path.exists() else self.image_path
+            return self._create_pixmap(path, size, dpr, is_svg=False)
+
+    def _create_pixmap(self, image_path, size, dpr, is_svg=False):
+        """Create a pixmap from image file with proper scaling and dark mode support."""
+        # Create high-resolution pixmap
+        scaled_size = QtCore.QSize(int(size.width() * dpr), int(size.height() * dpr))
+        pixmap = QtGui.QPixmap(scaled_size)
+        pixmap.setDevicePixelRatio(dpr)
+        pixmap.fill(QtCore.Qt.GlobalColor.transparent)
+
+        if is_svg:
+            return self._render_svg(image_path, pixmap, size, dpr)
+        else:
+            return self._render_png(image_path, pixmap, scaled_size, dpr)
```

Comment:
We can simpify this by merging `_create_pixmap`:

```suggestion
        scaled_size = QtCore.QSize(int(size.width() * dpr), int(size.height() * dpr))
        pixmap = QtGui.QPixmap(scaled_size)
        pixmap.setDevicePixelRatio(dpr)
        pixmap.fill(QtCore.Qt.GlobalColor.transparent)
        
        # Try SVG first, then fall back to PNG
        svg_path = self.image_path.with_suffix('.svg')
        if svg_path.exists():
            return self._render_svg(svg_path, pixmap, size, dpr)
        else:
            large_path = self.image_path.with_name(self.image_path.stem + '_large.png')
            path = large_path if large_path.exists() else self.image_path
            return self._render_png(path, pixmap, scaled_size, dpr)
```